core/types/subtyping.cc

```diff
@@ -404,21 +404,12 @@ TypePtr Types::lub(const GlobalState &gs, const TypePtr &t1, const TypePtr &t2)
             ENFORCE(result != nullptr);
             return result;
         } else {
-            bool allowProxyInLub = isa_type<TupleType>(t1) || isa_type<ShapeType>(t1);
             // only 1st is proxy
-            TypePtr und = t1.underlying(gs);
-            if (isSubType(gs, und, t2)) {
-                return t2;
-            } else if (allowProxyInLub) {
-                return OrType::make_shared(t1, t2);
-            } else {
-                return lub(gs, t2, und);
-            }
+            return lub(gs, t2, t1);
```

Comment:
This results in infinite recursion when we hit the logic up on line 182:

https://github.com/sorbet/sorbet/pull/4495/files#diff-eb2da3e78eea17626fb563370b444cc444ebc51aa7b8cd35da2fad4e43cbdd36R182-R184

Note also that even if this issue weren't present, all the logic in the runup to the `is_proxy_type(t1)` check would be repeated, which could have perf implications. So I think we might have to live with the repetition here, or come up with a different strategy to DRY it.

Reply:
Ah thanks :+1:, I suspected such a thing was in place as I was only seeing certain sequences and not others. I'll think about other ways to DRY this up.